### PR TITLE
Tech/create loading gif module

### PR DESCRIPTION
**Pull request by alschmut**
# Tech/create loading gif module

Merge permission: CC-Member

## Description

- Extract both loadingGifs from codeChartaController and RibbonBarController into a separate component `ui/loadingGif`

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

* [x] **All** requirements mentioned in the issue are implemented
* [x] Does match the Code of Conduct and the Contribution file 
* [ ] Task has its own **GitHub issue** (something it is solving)
  * [ ] Issue number is **included in the commit messages** for traceability
* [ ] **Update the README.md** with any changes/additions made
* [ ] **Update the CHANGELOG.md** with any changes/additions made
* [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
* [x] **All tests pass**    
* [x] **Descriptive pull request text**, answering:
  + What problem/issue are you fixing?
  + What does this PR implement and how? 
* [ ] **Assign your PR to someone** for a code review
  + This person _will be contacted **first**_ if a bug is introduced into `master`
* [x] **Manual testing** did not fail

### CODE REVIEW

File: visualization/app/codeCharta/codeCharta.component.html [hunk 1]
```diff
@@ -19,15 +19,13 @@
     <sidenav-toggle-component></sidenav-toggle-component>
 </div>
 
-<code-map-component ng-hide="$ctrl._viewModel.isLoadingFile"></code-map-component>
```
Comment (NearW) [inline visualization/app/codeCharta/codeCharta.component.html]: We should try to keep the map hidden again if a file is loading

File: visualization/app/codeCharta/ui/loadingGif/loadingGif.component.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        isLoadingFile: true,
+        isLoadingMap: true
+    }
+
+    /* @ngInject */
+    constructor(
+        private $rootScope: IRootScopeService,
+        private $timeout: ITimeoutService
+    ){
+        LoadingGifController.subscribe(this.$rootScope, this)
```
Comment (NearW) [inline visualization/app/codeCharta/ui/loadingGif/loadingGif.component.ts]: The LoadingGifSubscriber is only used here. In that case, we could also make methods such as "onLoadingFileStatusChanged" static and don't use angular events at all

File: visualization/app/codeCharta/ui/loadingGif/loadingGif.e2e.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import {CC_URL, puppeteer} from "../../../puppeteer.helper"
```
[the remark this answers is not in the document]
Comment (alschmut) [reply visualization/app/codeCharta/ui/loadingGif/loadingGif.e2e.ts]: It's auto generated by `npm run g` (plop generator), but you're right. It doesn't make sense to keep an empty test which runs for 1 second.

[the remark this answers is not in the document]
Comment (alschmut) [reply visualization/app/codeCharta/ui/loadingGif/loadingGif.e2e.ts]: What do you think is the correct implementation for the generator?

File: visualization/app/codeCharta/ui/loadingGif/loadingGif.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            rebuildController()
+
+            expect(LoadingGifController.subscribe).toHaveBeenCalledWith($rootScope, loadingGifController)
+        })
+
+        it("should set attribute isLoadingFile to true", () => {
+            rebuildController()
+
+            expect(loadingGifController["_viewModel"].isLoadingFile).toBeTruthy()
+            expect(loadingGifController["_viewModel"].isLoadingMap).toBeTruthy()
```
[the remark this answers is not in the document]
Comment (alschmut) [reply visualization/app/codeCharta/ui/loadingGif/loadingGif.component.spec.ts]: I will split this into two test cases

File: visualization/app/codeCharta/ui/loadingGif/loadingGif.component.ts (hunk 2 above)
Comment (alschmut) [reply visualization/app/codeCharta/ui/loadingGif/loadingGif.component.ts]: I actually prefer sending events from the other components than calling the loadingGif component directly, but if you have a good idea, tell me what you think of ;)

File: visualization/app/codeCharta/codeCharta.component.html (hunk 1 above)
Comment (alschmut) [reply visualization/app/codeCharta/codeCharta.component.html]: I implemented the `LoadingGifSubscriber` inside the `codeMap.component`

File: visualization/app/codeCharta/ui/loadingGif/loadingGif.e2e.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (NearW) [reply visualization/app/codeCharta/ui/loadingGif/loadingGif.e2e.ts]: Generator will generate an e2e tests, but if it's not adapted, it should be removed before committing in the future.

File: visualization/app/codeCharta/ui/loadingGif/loadingGif.component.ts (hunk 2 above)
Comment (NearW) [reply visualization/app/codeCharta/ui/loadingGif/loadingGif.component.ts]: now handled with a loadingGif.service

**Comment by alschmut** [comment]
@NearW I approve the new changes and we can merge this as soon as Travis is done ;)

**Pull request merged**